### PR TITLE
add project descriptor exclude and include support for build

### PR DESCRIPTION
**Pull request by elbandito**
Supports Project Descriptor (project.toml) `exclude` and `include` for `pack build` command. 
 
https://github.com/buildpacks/pack/issues/511

### CODE REVIEW

File: internal/archive/archive.go [hunk 1]
```diff
@@ -21,15 +21,15 @@ func init() {
 	NormalizedDateTime = time.Date(1980, time.January, 1, 0, 0, 1, 0, time.UTC)
 }
 
-func ReadDirAsTar(srcDir, basePath string, uid, gid int, mode int64, normalizeModTime bool) io.ReadCloser {
+func ReadDirAsTar(srcDir, basePath string, uid, gid int, mode int64, normalizeModTime bool, filter func(string) bool) io.ReadCloser {
```
Comment (jromero) [inline internal/archive/archive.go]: in this file, for readability, a type here would look nicer. Please we can document it.

File: internal/archive/archive_test.go [hunk 2]
```diff
@@ -100,7 +100,7 @@ func testArchive(t *testing.T, when spec.G, it spec.S) {
 
 				tw := tar.NewWriter(fh)
 
-				err = archive.WriteDirToTar(tw, src, "/nested/dir/dir-in-archive", 1234, 2345, 0777, true)
+				err = archive.WriteDirToTar(tw, src, "/nested/dir/dir-in-archive", 1234, 2345, 0777, true, nil)
```
Comment (jromero) [inline internal/archive/archive_test.go]: It's starting to look like we're got a decent amount of options here. Specially for optional args, we should probably move to  using `With` options pattern.

File: build.go [hunk 3]
```diff
@@ -48,6 +48,7 @@ type BuildOptions struct {
 	Buildpacks        []string
 	ProxyConfig       *ProxyConfig // defaults to  environment proxy vars
 	ContainerConfig   ContainerConfig
+	FileFilter        build.FileFilter
```
Comment (jromero) [inline build.go]: We should redefine this type here instead of depending on the use of an internal definition.

File: internal/archive/archive.go [hunk 4]
```diff
@@ -148,8 +148,14 @@ func ReadTarEntry(rc io.Reader, entryPath string) (*tar.Header, []byte, error) {
...
 // contents will be places.
-func WriteDirToTar(tw *tar.Writer, srcDir, basePath string, uid, gid int, mode int64, normalizeModTime bool) error {
+func WriteDirToTar(tw *tar.Writer, srcDir, basePath string, uid, gid int, mode int64, normalizeModTime bool, filter func(string) bool) error {
 	return filepath.Walk(srcDir, func(file string, fi os.FileInfo, err error) error {
+		if filter != nil && !filter(file) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
```
Comment (jromero) [inline internal/archive/archive.go]:
```suggestion
```

Duplicate error check

File: internal/archive/archive_test.go (hunk 2 above)
Comment (jromero) [reply internal/archive/archive_test.go]: We can opt to punt on this but I can see that only the first 3 args are required and all others should use `With` ops pattern and have sensible defaults.

File: build.go (hunk 3 above)
Comment (elbandito) [reply build.go]: @jromero when you say redefine are you saying:
```
FileFilter   func(string) bool
```
or add a new type in the file
```
type FileFilter func(string) bool
```

File: internal/archive/archive_test.go (hunk 2 above)
Comment (elbandito) [reply internal/archive/archive_test.go]: Sounds good.  Originally, I was looking for a way to perform method overloading, but unfortunately go doesn't support this.  This pattern seems like a good alternative.  I'll punt on this for this PR, but will include it into a separate refactor PR.

**Comment by jromero** [comment]
@elbandito code looks really good. Is this ready for review? It's currently in draft.

**Comment by elbandito** [comment]
> @elbandito code looks really good. Is this ready for review? It's currently in draft.

Yes, ready for review, thanks @jromero !

**Comment by jromero** [comment]
# ✅ User Accepted

### Excluding files and directories

```
$ tree ../samples/apps/bash-script/
../samples/apps/bash-script/
├── README.md
├── app.sh
├── bash-script-buildpack
│   ├── bin
│   │   ├── build
│   │   └── detect
│   └── buildpack.toml
└── project.toml

2 directories, 6 files

$ cat ../samples/apps/bash-script/project.toml 
[project]
id = "io.buildpacks.bash-script"
name = "Bash Script"
version = "1.0.0"

[build]
exclude = [
    "README.md",
    "bash-script-buildpack"
]

[[build.buildpacks]]
uri = "bash-script-buildpack/"

$./out/pack build bash-script -B cnbs/sample-builder:alpine -p ../samples/apps/bash-script/ -v
Using project descriptor located at ../samples/apps/bash-script/project.toml
Pulling image index.docker.io/cnbs/sample-builder:alpine
alpine: Pulling from cnbs/sample-builder
Digest: sha256:f199c03ebe56d78c3a332591e9a861bd83bfbf671cd9f01308dcce74ea2093de
Status: Image is up to date for cnbs/sample-builder:alpine
Selected run image cnbs/sample-stack-run:alpine
Pulling image cnbs/sample-stack-run:alpine
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:fa05d2d97a4d7b18ddd338f093f309b6fdaff71931675544ec0fe0fdb8d90fb6
Status: Image is up to date for cnbs/sample-stack-run:alpine
Adding buildpack io.buildpacks.samples.bash-script version 0.0.1 to builder
Setting custom order
Using build cache volume pack-cache-47a4611cdf37.build
===> DETECTING
[detector] ======== Output: io.buildpacks.samples.bash-script@0.0.1 ========
[detector] ---> Hello Bash Script buildpack
[detector] 
[detector] Here are the contents of the current working directory:
[detector] .:
[detector] total 16
[detector] drwxr-xr-x    2 cnb      cnb           4096 Apr  2 17:16 .
[detector] drwxr-xr-x    1 root     root          4096 Apr  2 17:16 ..
[detector] -rwxr-xr-x    1 cnb      cnb            738 Apr  2 16:57 app.sh
[detector] -rw-r--r--    1 cnb      cnb            202 Apr  2 16:37 project.toml
[detector] ======== Results ========
[detector] pass: io.buildpacks.samples.bash-script@0.0.1
[detector] Resolving plan... (try #1)
[detector] io.buildpacks.samples.bash-script 0.0.1
===> ANALYZING
[analyzer] Analyzing image "a017655439e75ab106657810960f07ef0d5ee4ff691a7e7423674d5f666d2366"
===> RESTORING
===> BUILDING
[builder] ---> Hello World buildpack
[builder] ---> Hello Bash Script buildpack
[builder] 
[builder] Here are the contents of the current working directory:
[builder] .:
[builder] total 16
[builder] drwxr-xr-x    2 cnb      cnb           4096 Apr  2 17:16 .
[builder] drwxr-xr-x    1 root     root          4096 Apr  2 17:16 ..
[builder] -rwxr-xr-x    1 cnb      cnb            738 Apr  2 16:57 app.sh
[builder] -rw-r--r--    1 cnb      cnb            202 Apr  2 16:37 project.toml
===> EXPORTING
[exporter] Reusing layers from image with id 'a017655439e75ab106657810960f07ef0d5ee4ff691a7e7423674d5f666d2366'
[exporter] Writing tarball for layer "launcher"
[exporter] Reusing layer 'launcher'
[exporter] Layer 'launcher' SHA: sha256:6fb83b26bebe88ae50154b591d1272397d02e5ba679a55278ed756ff7f94d526
[exporter] Layer 'app' SHA: sha256:4c6ff16e28deaf437a1abe5c926cdfb8afcd2c31ea544f6992ed7b665965b3f4
[exporter] Reusing 1/1 app layer(s)
[exporter] Writing tarball for layer "config"
[exporter] Reusing layer 'config'
[exporter] Layer 'config' SHA: sha256:fe7742acb11ad04073b715af45ed319d84f74a811bc9a0f1245379391c62aab5
[exporter] *** Images (a017655439e7):
[exporter]       index.docker.io/library/bash-script:latest
[exporter] 
[exporter] *** Image ID: a017655439e75ab106657810960f07ef0d5ee4ff691a7e7423674d5f666d2366
Successfully built image bash-script

$ docker run -it bash-script

    |'-_ _-'|       ____          _  _      _                      _             _
    |   |   |      |  _ \        (_)| |    | |                    | |           (_)
     '-_|_-'       | |_) | _   _  _ | |  __| | _ __    __ _   ___ | | __ ___     _   ___
|'-_ _-'|'-_ _-'|  |  _ < | | | || || | / _` || '_ \  / _` | / __|| |/ // __|   | | / _ \
|   |   |   |   |  | |_) || |_| || || || (_| || |_) || (_| || (__ |   < \__ \ _ | || (_) |
 '-_|_-' '-_|_-'   |____/  \__,_||_||_| \__,_|| .__/  \__,_| \___||_|\_\|___/(_)|_| \___/
                                              | |
                                              |_|


Here are the contents of the current working directory:
.:
total 16
drwxr-xr-x    2 cnb      cnb           4096 Jan  1  1980 .
drwxr-xr-x    1 root     root          4096 Apr  2 17:17 ..
-rwxr-xr-x    1 cnb      cnb            738 Jan  1  1980 app.sh
-rw-r--r--    1 cnb      cnb            202 Jan  1  1980 project.toml

```

### Include/Exclude are mutually exclusive

```
$ tree ../samples/apps/bash-script/
../samples/apps/bash-script/
├── README.md
├── app.sh
├── bash-script-buildpack
│   ├── bin
│   │   ├── build
│   │   └── detect
│   └── buildpack.toml
└── project.toml

2 directories, 6 files

$ cat ../samples/apps/bash-script/project.toml 
[project]
id = "io.buildpacks.bash-script"
name = "Bash Script"
version = "1.0.0"

[build]
include = [
    "README.md"
]
exclude = [
    "README.md",
    "bash-script-buildpack"
]

[[build.buildpacks]]
uri = "bash-script-buildpack/"

$ ./out/pack build bash-script -B cnbs/sample-builder:alpine -p ../samples/apps/bash-script/ -v
ERROR: project.toml: cannot have both include and exclude defined
```

### Include excludes everything else

```
$ tree ../samples/apps/bash-script/
../samples/apps/bash-script/
├── README.md
├── app.sh
├── bash-script-buildpack
│   ├── bin
│   │   ├── build
│   │   └── detect
│   └── buildpack.toml
└── project.toml

2 directories, 6 files

$ cat ../samples/apps/bash-script/project.toml 
[project]
id = "io.buildpacks.bash-script"
name = "Bash Script"
version = "1.0.0"

[build]
include = [
    "app.sh"
]

[[build.buildpacks]]
uri = "bash-script-buildpack/"

$ ./out/pack build bash-script -B cnbs/sample-builder:alpine -p ../samples/apps/bash-script/ -v
Using project descriptor located at ../samples/apps/bash-script/project.toml
Pulling image index.docker.io/cnbs/sample-builder:alpine
alpine: Pulling from cnbs/sample-builder
Digest: sha256:f199c03ebe56d78c3a332591e9a861bd83bfbf671cd9f01308dcce74ea2093de
Status: Image is up to date for cnbs/sample-builder:alpine
Selected run image cnbs/sample-stack-run:alpine
Pulling image cnbs/sample-stack-run:alpine
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:fa05d2d97a4d7b18ddd338f093f309b6fdaff71931675544ec0fe0fdb8d90fb6
Status: Image is up to date for cnbs/sample-stack-run:alpine
Adding buildpack io.buildpacks.samples.bash-script version 0.0.1 to builder
Setting custom order
Using build cache volume pack-cache-47a4611cdf37.build
===> DETECTING
[detector] ======== Output: io.buildpacks.samples.bash-script@0.0.1 ========
[detector] ---> Hello Bash Script buildpack
[detector] 
[detector] Here are the contents of the current working directory:
[detector] .:
[detector] total 12
[detector] drwxr-xr-x    2 cnb      cnb           4096 Apr  2 17:22 .
[detector] drwxr-xr-x    1 root     root          4096 Apr  2 17:22 ..
[detector] -rwxr-xr-x    1 cnb      cnb            738 Apr  2 16:57 app.sh
[detector] ======== Results ========
[detector] pass: io.buildpacks.samples.bash-script@0.0.1
[detector] Resolving plan... (try #1)
[detector] io.buildpacks.samples.bash-script 0.0.1
===> ANALYZING
[analyzer] Analyzing image "a017655439e75ab106657810960f07ef0d5ee4ff691a7e7423674d5f666d2366"
===> RESTORING
===> BUILDING
[builder] ---> Hello World buildpack
[builder] ---> Hello Bash Script buildpack
[builder] 
[builder] Here are the contents of the current working directory:
[builder] .:
[builder] total 12
[builder] drwxr-xr-x    2 cnb      cnb           4096 Apr  2 17:22 .
[builder] drwxr-xr-x    1 root     root          4096 Apr  2 17:22 ..
[builder] -rwxr-xr-x    1 cnb      cnb            738 Apr  2 16:57 app.sh
===> EXPORTING
[exporter] Reusing layers from image with id 'a017655439e75ab106657810960f07ef0d5ee4ff691a7e7423674d5f666d2366'
[exporter] Writing tarball for layer "launcher"
[exporter] Reusing layer 'launcher'
[exporter] Layer 'launcher' SHA: sha256:6fb83b26bebe88ae50154b591d1272397d02e5ba679a55278ed756ff7f94d526
[exporter] Layer 'app' SHA: sha256:b902cec8583796fcab3827b84b310dd84a53c4b788d5d002fd06de4581ec6664
[exporter] Adding 1/1 app layer(s)
[exporter] Writing tarball for layer "config"
[exporter] Reusing layer 'config'
[exporter] Layer 'config' SHA: sha256:fe7742acb11ad04073b715af45ed319d84f74a811bc9a0f1245379391c62aab5
[exporter] *** Images (bad3068e5afa):
[exporter]       index.docker.io/library/bash-script:latest
[exporter] 
[exporter] *** Image ID: bad3068e5afaf64fa61e7382f492ff989e493045e6921169aa85c9a93b2a9437
Successfully built image bash-script

$ docker run -it bash-script

    |'-_ _-'|       ____          _  _      _                      _             _
    |   |   |      |  _ \        (_)| |    | |                    | |           (_)
     '-_|_-'       | |_) | _   _  _ | |  __| | _ __    __ _   ___ | | __ ___     _   ___
|'-_ _-'|'-_ _-'|  |  _ < | | | || || | / _` || '_ \  / _` | / __|| |/ // __|   | | / _ \
|   |   |   |   |  | |_) || |_| || || || (_| || |_) || (_| || (__ |   < \__ \ _ | || (_) |
 '-_|_-' '-_|_-'   |____/  \__,_||_||_| \__,_|| .__/  \__,_| \___||_|\_\|___/(_)|_| \___/
                                              | |
                                              |_|


Here are the contents of the current working directory:
.:
total 12
drwxr-xr-x    2 cnb      cnb           4096 Jan  1  1980 .
drwxr-xr-x    1 root     root          4096 Apr  2 17:22 ..
-rwxr-xr-x    1 cnb      cnb            738 Jan  1  1980 app.sh
```

### Glob

```
$ tree ../samples/apps/bash-script/
../samples/apps/bash-script/
├── README.md
├── app.sh
├── bash-script-buildpack
│   ├── bin
│   │   ├── build
│   │   └── detect
│   └── buildpack.toml
└── project.toml

2 directories, 6 files

$ cat ../samples/apps/bash-script/project.toml 
[project]
id = "io.buildpacks.bash-script"
name = "Bash Script"
version = "1.0.0"

[build]
exclude = [
    "README.md",
    "bash-script-buildpack/*/build"
]

[[build.buildpacks]]
uri = "bash-script-buildpack/"

$ ./out/pack build bash-script -B cnbs/sample-builder:alpine -p ../samples/apps/bash-script/ -v
Using project descriptor located at ../samples/apps/bash-script/project.toml
Pulling image index.docker.io/cnbs/sample-builder:alpine
alpine: Pulling from cnbs/sample-builder
Digest: sha256:f199c03ebe56d78c3a332591e9a861bd83bfbf671cd9f01308dcce74ea2093de
Status: Image is up to date for cnbs/sample-builder:alpine
Selected run image cnbs/sample-stack-run:alpine
Pulling image cnbs/sample-stack-run:alpine
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:fa05d2d97a4d7b18ddd338f093f309b6fdaff71931675544ec0fe0fdb8d90fb6
Status: Image is up to date for cnbs/sample-stack-run:alpine
Adding buildpack io.buildpacks.samples.bash-script version 0.0.1 to builder
Setting custom order
Using build cache volume pack-cache-47a4611cdf37.build
===> DETECTING
[detector] ======== Output: io.buildpacks.samples.bash-script@0.0.1 ========
[detector] ---> Hello Bash Script buildpack
[detector] 
[detector] Here are the contents of the current working directory:
[detector] .:
[detector] total 20
[detector] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 17:28 .
[detector] drwxr-xr-x    1 root     root          4096 Apr  2 17:28 ..
[detector] -rwxr-xr-x    1 cnb      cnb            738 Apr  2 16:57 app.sh
[detector] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 16:10 bash-script-buildpack
[detector] -rw-r--r--    1 cnb      cnb            210 Apr  2 17:28 project.toml
[detector] 
[detector] ./bash-script-buildpack:
[detector] total 16
[detector] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 16:10 .
[detector] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 17:28 ..
[detector] drwxr-xr-x    2 cnb      cnb           4096 Apr  2 16:26 bin
[detector] -rw-r--r--    1 cnb      cnb            320 Apr  2 16:10 buildpack.toml
[detector] 
[detector] ./bash-script-buildpack/bin:
[detector] total 12
[detector] drwxr-xr-x    2 cnb      cnb           4096 Apr  2 16:26 .
[detector] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 16:10 ..
[detector] -rwxr-xr-x    1 cnb      cnb            242 Apr  2 16:20 detect
[detector] ======== Results ========
[detector] pass: io.buildpacks.samples.bash-script@0.0.1
[detector] Resolving plan... (try #1)
[detector] io.buildpacks.samples.bash-script 0.0.1
===> ANALYZING
[analyzer] Analyzing image "bad3068e5afaf64fa61e7382f492ff989e493045e6921169aa85c9a93b2a9437"
===> RESTORING
===> BUILDING
[builder] ---> Hello World buildpack
[builder] ---> Hello Bash Script buildpack
[builder] 
[builder] Here are the contents of the current working directory:
[builder] .:
[builder] total 20
[builder] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 17:28 .
[builder] drwxr-xr-x    1 root     root          4096 Apr  2 17:28 ..
[builder] -rwxr-xr-x    1 cnb      cnb            738 Apr  2 16:57 app.sh
[builder] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 16:10 bash-script-buildpack
[builder] -rw-r--r--    1 cnb      cnb            210 Apr  2 17:28 project.toml
[builder] 
[builder] ./bash-script-buildpack:
[builder] total 16
[builder] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 16:10 .
[builder] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 17:28 ..
[builder] drwxr-xr-x    2 cnb      cnb           4096 Apr  2 16:26 bin
[builder] -rw-r--r--    1 cnb      cnb            320 Apr  2 16:10 buildpack.toml
[builder] 
[builder] ./bash-script-buildpack/bin:
[builder] total 12
[builder] drwxr-xr-x    2 cnb      cnb           4096 Apr  2 16:26 .
[builder] drwxr-xr-x    3 cnb      cnb           4096 Apr  2 16:10 ..
[builder] -rwxr-xr-x    1 cnb      cnb            242 Apr  2 16:20 detect
===> EXPORTING
[exporter] Reusing layers from image with id 'bad3068e5afaf64fa61e7382f492ff989e493045e6921169aa85c9a93b2a9437'
[exporter] Writing tarball for layer "launcher"
[exporter] Reusing layer 'launcher'
[exporter] Layer 'launcher' SHA: sha256:6fb83b26bebe88ae50154b591d1272397d02e5ba679a55278ed756ff7f94d526
[exporter] Layer 'app' SHA: sha256:81663360a6eb0ce8b32559bc9e1d1c339ddc03ba3f99c33bb2022ccafafac7e4
[exporter] Adding 1/1 app layer(s)
[exporter] Writing tarball for layer "config"
[exporter] Reusing layer 'config'
[exporter] Layer 'config' SHA: sha256:fe7742acb11ad04073b715af45ed319d84f74a811bc9a0f1245379391c62aab5
[exporter] *** Images (3723de85412a):
[exporter]       index.docker.io/library/bash-script:latest
[exporter] 
[exporter] *** Image ID: 3723de85412a53cf0e546e56c32247d12d70365fcabf14f646bac6d8932e0741
Successfully built image bash-script


$ docker run -it bash-script

    |'-_ _-'|       ____          _  _      _                      _             _
    |   |   |      |  _ \        (_)| |    | |                    | |           (_)
     '-_|_-'       | |_) | _   _  _ | |  __| | _ __    __ _   ___ | | __ ___     _   ___
|'-_ _-'|'-_ _-'|  |  _ < | | | || || | / _` || '_ \  / _` | / __|| |/ // __|   | | / _ \
|   |   |   |   |  | |_) || |_| || || || (_| || |_) || (_| || (__ |   < \__ \ _ | || (_) |
 '-_|_-' '-_|_-'   |____/  \__,_||_||_| \__,_|| .__/  \__,_| \___||_|\_\|___/(_)|_| \___/
                                              | |
                                              |_|


Here are the contents of the current working directory:
.:
total 20
drwxr-xr-x    3 cnb      cnb           4096 Jan  1  1980 .
drwxr-xr-x    1 root     root          4096 Apr  2 17:29 ..
-rwxr-xr-x    1 cnb      cnb            738 Jan  1  1980 app.sh
drwxr-xr-x    3 cnb      cnb           4096 Jan  1  1980 bash-script-buildpack
-rw-r--r--    1 cnb      cnb            210 Jan  1  1980 project.toml

./bash-script-buildpack:
total 16
drwxr-xr-x    3 cnb      cnb           4096 Jan  1  1980 .
drwxr-xr-x    3 cnb      cnb           4096 Jan  1  1980 ..
drwxr-xr-x    2 cnb      cnb           4096 Jan  1  1980 bin
-rw-r--r--    1 cnb      cnb            320 Jan  1  1980 buildpack.toml

./bash-script-buildpack/bin:
total 12
drwxr-xr-x    2 cnb      cnb           4096 Jan  1  1980 .
drwxr-xr-x    3 cnb      cnb           4096 Jan  1  1980 ..
-rwxr-xr-x    1 cnb      cnb            242 Jan  1  1980 detect
```

**Comment by jromero** [comment]
Some final comments/finding while running through acceptance. We will carry these over into potentially further improving this feature:

* In exclude, an entry for directory with trailing file path separator will be retained `buildpacks/` but contents excluded.
* Should `include` infer `project.toml` as included? Currently, it is excluded just as any other content would be.
* If we use .gitignore semantics couldn’t we collapse include and exclude into one?

CC: @jkutner / @elbandito

**Comment by elbandito** [comment]
> * If we use .gitignore semantics couldn’t we collapse include and exclude into one?
> 
We could potentially do something similar as `.gitignore`.  For example if we wanted to `exclude` all jar files expect the `app.jar` we could have something like: 
```
[build]
ignore = [ 
    "*.jar", 
    "!app.jar"
]
```
It's an interesting idea that I'd be curious to hear other's thoughts on.

**Pull request merged**